README.md
# middy-middleware-json-error-handler

[![npm version](https://badge.fury.io/js/middy-middleware-json-error-handler.svg)](https://npmjs.org/package/middy-middleware-json-error-handler)
[![downloads](https://img.shields.io/npm/dw/middy-middleware-json-error-handler.svg)](https://npmjs.org/package/middy-middleware-json-error-handler)
[![open issues](https://img.shields.io/github/issues-raw/dbartholomae/middy-middleware-json-error-handler.svg)](https://github.com/dbartholomae/middy-middleware-json-error-handler/issues)
[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2Fdbartholomae%2Fmiddy-middleware-json-error-handler.svg?type=shield)](https://app.fossa.io/projects/git%2Bgithub.com%2Fdbartholomae%2Fmiddy-middleware-json-error-handler?ref=badge_shield)
[![debug](https://img.shields.io/badge/debug-blue.svg)](https://github.com/visionmedia/debug#readme)
[![build status](https://github.com/dbartholomae/middy-middleware-json-error-handler/workflows/Build%20and%20deploy/badge.svg?branch=main)](https://github.com/dbartholomae/middy-middleware-json-error-handler/actions?query=workflow%3A"Build+and+deploy")
[![codecov](https://codecov.io/gh/dbartholomae/middy-middleware-json-error-handler/branch/main/graph/badge.svg)](https://codecov.io/gh/dbartholomae/middy-middleware-json-error-handler)
[![semantic release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release#badge)

A [middy](https://github.com/middyjs/middy) middleware that returns errors as http errors, compatible with [http-errors](https://www.npmjs.com/package/http-errors).

## Installation
Download node at [nodejs.org](http://nodejs.org) and install it, if you haven't already.

```sh
npm install middy-middleware-json-error-handler --save
```

## Documentation

There is [additional documentation](https://dbartholomae.github.com/middy-middleware-json-error-handler).

## Usage

```typescript
import middy from '@middy/core'
import JSONErrorHandlerMiddleware from 'middy-middleware-json-error-handler'
import createHttpError from 'http-errors'
import { APIGatewayEvent } from 'aws-lambda'

// This is your AWS handler
async function helloWorld (event: APIGatewayEvent) {
  if (event.queryStringParameters?.search == null) {
    // If you throw an error with status code, the error will be returned as stringified JSON.
    // Only the stack will be omitted.
    throw createHttpError(400, 'Query has to include a search')
  }

  if (event.queryStringParameters?.search === 'error') {
    // If you throw an error with status code greater than 500 and specify in options that the error
    // must be exposed, then the error will be returned as stringified JSON
    throw createHttpError(500, 'Something went wrong', { expose: true })
  }

  // If you throw an error with no status code, only a generic message will be shown to the user
  // instead of the full error
  throw new Error('Search is not implemented yet')
}

// Specify which are the error properties to omit in response (by default, only the stack is omitted)
const options = {
  errorPropertiesToOmit: ['name', 'statusCode', 'stack']
}

// Let's "middyfy" our handler, then we will be able to attach middlewares to it (options parameter is optional)
export const handler = middy(helloWorld)
  .use(JSONErrorHandlerMiddleware(options)) // This middleware is needed do handle the errors thrown by the handler
```


CHANGELOG.md
# [4.0.0](https://github.com/dbartholomae/middy-middleware-json-error-handler/compare/3.1.0...4.0.0) (2023-07-26)


### Features

* allow to expose internal errors via expose property ([309084d](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/309084dc6067400fd915200f69aac1a3b9ff140d))


### BREAKING CHANGES

* errors with the `expose` property set to a truthy
value will be exposed even if their status code is >= 500

# [3.1.0](https://github.com/dbartholomae/middy-middleware-json-error-handler/compare/3.0.1...3.1.0) (2023-07-22)


### Features

* allow choosing which error properties to omit ([d395b92](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/d395b920aa2382e749ff7673f040cf700d8a7876))

## [3.0.1](https://github.com/dbartholomae/middy-middleware-json-error-handler/compare/3.0.0...3.0.1) (2023-05-14)


### Bug Fixes

* update dependencies that show security warnings ([e3ba58c](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/e3ba58cbb8f29b610abe2739a81b62db2454805b))

# [3.0.0](https://github.com/dbartholomae/middy-middleware-json-error-handler/compare/2.0.6...3.0.0) (2021-08-01)


### Features

* update all dependencies ([7c85db9](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/7c85db93a46a491e21d980332f10af7c61947151))


### BREAKING CHANGES

* The middleware is now using the new types from middy 2. Based on a quick test, without TypeScript, the new version is still compatible with middy 1.

## [2.0.6](https://github.com/dbartholomae/middy-middleware-json-error-handler/compare/2.0.5...2.0.6) (2020-08-30)


### Bug Fixes

* remove snyk ([61674b8](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/61674b87fc4a7b897efa0d624fd4a5d0117215b9))
* update all dependencies ([70f1539](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/70f15397b3f3b6f175755df15ad3cb029c243213))

## [2.0.5](https://github.com/dbartholomae/middy-middleware-json-error-handler/compare/2.0.4...2.0.5) (2020-08-17)


### Bug Fixes

* package.json to reduce vulnerabilities ([021cd9d](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/021cd9d0a94c367e2766c7474dddd9de8b3790dd))

## [2.0.4](https://github.com/dbartholomae/middy-middleware-json-error-handler/compare/2.0.3...2.0.4) (2020-08-09)


### Bug Fixes

* package.json to reduce vulnerabilities ([83e2653](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/83e2653fd8a9d1e4d6db70223f1f793b1d24d89a))

## [2.0.3](https://github.com/dbartholomae/middy-middleware-json-error-handler/compare/2.0.2...2.0.3) (2020-07-15)


### Bug Fixes

* depend on @types/debug only for development ([cbb0238](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/cbb02381c7574134f413506744aab2c3d4ffaf4f))

## [2.0.2](https://github.com/dbartholomae/middy-middleware-json-error-handler/compare/2.0.1...2.0.2) (2020-06-20)


### Bug Fixes

* package.json to reduce vulnerabilities ([5cb8437](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/5cb8437fbc0043759eb32f7f20494412e45f93cd))

## [2.0.1](https://github.com/dbartholomae/middy-middleware-json-error-handler/compare/2.0.0...2.0.1) (2020-05-19)


### Bug Fixes

* package.json & .snyk to reduce vulnerabilities ([dd0a096](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/dd0a096b672d46ad54e7f7ad0ee52ce626f50684))
* package.json & .snyk to reduce vulnerabilities ([da737ad](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/da737adf920dbd5110dcf46de6d87256527fb140))

# [2.0.0](https://github.com/dbartholomae/middy-middleware-json-error-handler/compare/1.0.3...2.0.0) (2020-05-02)


### Features

* update to middy 1.0.0 ([b536c03](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/b536c0310c2119f9d67bb19d03982353efc59d07))


### BREAKING CHANGES

* As of this version this middleware only supports middy@1.0.0 and upwards. See https://github.com/middyjs/middy/blob/master/UPGRADE.md for middy upgrade instructions.

## [1.0.3](https://github.com/dbartholomae/middy-middleware-json-error-handler/compare/1.0.2...1.0.3) (2020-05-01)


### Bug Fixes

* package.json & .snyk to reduce vulnerabilities ([ce3758b](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/ce3758be22d03567f5dfbfbe1da46a91ce502591))
* package.json & .snyk to reduce vulnerabilities ([57a433b](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/57a433b3ef1bd24814dbdb5a2cf6b4a8a821a8db))

## [1.0.2](https://github.com/dbartholomae/middy-middleware-json-error-handler/compare/1.0.1...1.0.2) (2020-04-17)


### Bug Fixes

* upgrade middy from 0.34.0 to 0.36.0 ([3af9ca9](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/3af9ca995a6bacb2ab32fa27f8a0ed6fbb3eb5ab))

## [1.0.1](https://github.com/dbartholomae/middy-middleware-json-error-handler/compare/1.0.0...1.0.1) (2020-04-10)


### Bug Fixes

* package.json & .snyk to reduce vulnerabilities ([f01fa53](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/f01fa5317068b7e2a5b6f37c8dbaa6c8905a3695))

# 1.0.0 (2020-01-23)


### Bug Fixes

* build with node 12 ([37ca2fe](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/37ca2fee71459d9c16de5797a2606f13fb819a43))


### Features

* add IErrorWithStatusCode interface and type guard ([0e0a90b](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/0e0a90bd06fced3d620a003e7d81419ddbba4d89))
* add logging for debugging ([f780ba2](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/f780ba22fe89ee02b3af0d03ea99be9e9c917f54))
* add omit helper function ([706696d](https://github.com/dbartholomae/middy-middleware-json-error-handler/commit/706696d4c84b6a15d31916f753d7de9da66e6cb7))

examples/helloWorld.ts
import { APIGatewayEvent } from 'aws-lambda'
import JSONErrorHandlerMiddleware from '../'
import createHttpError from 'http-errors'
import middy from '@middy/core'

// This is your AWS handler
async function helloWorld (event: APIGatewayEvent) {
  if (event.queryStringParameters?.search == null) {
    // If you throw an error with status code, the error will be returned as stringified JSON
    // Only the stack will be omitted.
    throw createHttpError(400, 'Query has to include a search')
  }

  if (event.queryStringParameters?.search === 'error') {
    // If you throw an error with status code greater than 500 and specify in options that the error
    // must be exposed, then the error will be returned as stringified JSON
    throw createHttpError(500, 'Something went wrong', { expose: true })
  }

  // If you throw an error with no status code, only a generic message will be shown to the user
  // instead of the full error
  throw new Error('Search is not implemented yet')
}

// Specify which are the error properties to omit in response (by default, only the stack is omitted)
const options = {
  errorPropertiesToOmit: ['name', 'statusCode', 'stack']
}

// Let's "middyfy" our handler, then we will be able to attach middlewares to it
export const handler = middy(helloWorld)
  .use(JSONErrorHandlerMiddleware(options)) // This middleware is needed do handle the errors thrown by the JWTAuthMiddleware
